staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -396,15 +396,31 @@ func (mounter *SafeFormatAndMount) formatAndMountSensitive(source string, target
 		}
 
 		// Disk is unformatted so format it.
-		args := []string{source}
+		args := []string{}
 		if fstype == "ext4" || fstype == "ext3" {
 			args = []string{
 				"-F",  // Force flag
 				"-m0", // Zero blocks reserved for super-user
-				source,
 			}
 		}
 
+		if fstype == "ext4" {
+			args = append(
+				args,
+				"-E",
+				"nodiscard",
```

Comment:
cc @jsafrane @gnufied 

I wonder if we need to expose this as an option to the function. This function is shared by almost all drivers and this is changing the default linux setting.

According to the manpage, "discard" is supposed to actually improve fs initialization? 

https://linux.die.net/man/8/mkfs.ext4
```
discard
Attempt to discard blocks at mkfs time (discarding blocks initially is useful on solid state devices and sparse / thin-provisioned storage). When the device advertises that discard also zeroes data (any subsequent read after the discard and before write returns zero), then mark all not-yet-zeroed inode tables as zeroed. This significantly speeds up filesystem initialization. This is set as default.
```

Reply:
It seems like it's very vendor-dependent:
https://wiki.debian.org/SSDOptimization#WARNING
https://unix.stackexchange.com/questions/366949/when-and-where-to-use-rw-nofail-noatime-discard-defaults
https://unix.stackexchange.com/questions/344800/mkfs-is-extremely-slow

This random site implies that while this may speed up fs creation, there may be a performance hit on deletion later. https://freethreads.weebly.com/ceph-development/quick-mkfs-on-a-big-device

Reply:
Hmm I have thought this is only relevant when a drive has been used and we are using `mkfs` to re-format it, maybe that is not the case?

Reply:
For local volumes, it's possible that disks get reused. For other types I'm not sure, as the man page also recommends using discard for thin-provisioned volumes. I think it's going to depend on how each provider exposes and reacts to the TRIM command.

Reply:
it only affects the drivers/users which explicitly set `MkfsNoDiscard` right? so it does not change default Linux settings by default as such?

Reply:
Also cc @sandeen .

Reply:
Yeah I believe so, the default value for bool is false IIUC

Reply:
From a very quick look, I have concerns that this conflates mkfs-time discard with runtime discard (I see code that appears to add the "discard" mount option as a new thing?)

mkfs-time discard of a fresh volume and runtime discard are not interchangeable behaviors.

Reply:
Thanks for taking a look @sandeen. The underlying reason we are trying to make this change is - we have observed that somtimes `mkfs.xfs` just hangs while formatting new volumes. There is a redhat portal article which points to - https://access.redhat.com/solutions/727333 and recommends folks to use `-K`  (and equivalent for ext4).

Reading your response it seems like - a provider may support runtime `discard` and yet may not support mkfs-time discard. But then @leiyiz  and @jingxu97 shouldn't this happen all the time and none of the XFS e2e tests would pass on GCE-PD? 

@sandeen we are trying to determine if it is safe to use `-K` or `-E discard` during `mkfs` time. I don't suppose there is a way we can query mkfs-time discard support from underlying volume?

Reply:
@sandeen yes they are not interchangeable, however, I have gotten confirmation from Persistent Disk team internally that as long as the disk is new, and the discard is passed at runtime, it is ok to use nodiscard at mkfs time. For other providers we are not sure that is why by default we keep the mkfs time behavior as is.

@gnufied does that answer your questions as well

Reply:
Again: passing discard at mount time is not interchangeable with or a substitute for using discard at mkfs time. It is a different mode of operation with different results.

You may or may not benefit from runtime discard, but it is orthogonal to the question of whole-device discard at mkfs time, and the two should not be conflated.

Reply:
Sorry I have miss-communicated what I meant so let me try to rephrase: for Persistent Disk specifically, when a disk is freshly created - which seems like is the case due to `existingFormat == ""`? - it doesn't matter if we use `nodiscard` at mkfs time.

Reply:
I'll try to clarify this one more time.

Discard at mkfs time discards the entire device prior to the format, i.e. it notifies the storage that NO blocks are currently in use, and then writes the format to the disk (effectively using some blocks from the just-discarded ranges).  If this is hanging for you, you have an _underlying storage bug_ which IMHO should not be ignored or worked around. It at least warrants investigation and understanding.

Mounting with "-o discard" will only discard blocks when they are freed by the filesystem, i.e. when you unlink a file. Any blocks which were unused at the initial mkfs time will remain un-discarded until they get allocated, written to, and then removed. If this doesn't happen they will remain un-discarded indefinitely. Further, runtime "-o discard" has possible performance implications, because you'ren ow adding a discard operation every time you free a block.

> we're adding an extra option in the mounter and existing code can keep functioning without behavioral change.

I want to be 100% clear here: changing from mkfs-time discard to "mount -o discard" absolutely _is_ a behavioral change.

Reply:
thanks for the explanation. I'd like to respond to your concerns:
1. we're only removing mkfs time discard for gce-pd, not for other providers. For them no behavior is changed
2. we're NOT "moving" mkfs time discard to mount time, we're just not gonna discard at mkfs time since the disk would not have been used before if [`existingFormat == ""`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/mount-utils/mount_linux.go#L428). (please advise if that's not the case)
3. we were adding mount time discard per recommendation of PD team but now we are not going to discard at mount time and leaving it for users to decide.
4. as for the underlying storage bug, I do agree but it seems to be an [existing issue](https://access.redhat.com/solutions/727333) not only for us but for linux community in general. So I believe it's not a issue we have enough expertise to investigate into.

Reply:
I think what Eric is saying is - it is not safe to workaround the hanging bug by using `-o nodiscard` during mkfs time. The linked article - https://access.redhat.com/solutions/727333 points at *one* root cause but that is not the only root cause and should be taken with a pinch of salt. In GCE-PD since `mkfs.xfs` most time work without hanging - the root cause is not same as https://access.redhat.com/solutions/727333 but something in storage provider. It would be wrong to workaround a bug in storage provider (GCE-PD in this case) in k8s. It could have unintended side effects and hence we should not purse this design.

Reply:
> thanks for the explanation. I'd like to respond to your concerns:
...
>     3. we were adding mount time discard per recommendation of PD team but now we are NOT going to discard at mount time and leaving it for users to decide, as is before.

Ok, I feel much better about it then. The initial explanation seemed to assume that one was a functional replacement for the other and I just wanted to drive home the fact that it is not.  Thanks!